### PR TITLE
T-32073 - Conditional property for join_distribution_type = PARTITIONED

### PR DESCRIPTION
**Pull request by belen-pruvost**
Follow up from [this discussion](https://dune.height.app/T-32073#e2b19a18-daac-4412-bd54-b4537097a994)

### CODE REVIEW

**Comment by belen-pruvost** [comment]
I believe this model will likely cause all models to be affected

File: dbt_project.yml
Comment (jkylling) [inline dbt_project.yml:53]: Nit: Could we make this more flexible and allow for configuring any join distribution type on the model. E.g.,
```
{{ set_trino_session_property(model.config.get('join_distribution_type', false), 'join_distribution_type', model.config.get('join_distribution_type'))) }}
```

Comment (jkylling) [reply dbt_project.yml:53]: Or add a new macro to support a simpler syntax.

Comment (belen-pruvost) [reply dbt_project.yml]: Yeah - this makes more sense 👍 I've opened this property to all of the property's possible values.

**Comment by jeff-dude** [comment]
manually canceling CI, as it kicks off large subset (should do that each commit on these types of PRs)

File: macros/set_trino_session_property.sql
Comment (belen-pruvost) [reply macros/set_trino_session_property.sql:13]: We leave the option open for the future to use different partitioning types.
However, for now, you would just need to set `join_distribution_type='PARTITIONED'` to the models that result into `EXCEEDED_LOCAL_MEMORY_LIMIT` (like `nft_polygon.transfers`)

**Comment by belen-pruvost** [comment]
> manually canceling CI, as it kicks off large subset (should do that each commit on these types of PRs)

I still need to test if this is setting the session property as expected, I will limit the model run step to just this model then. Are there other options otherwise?

And besides that, for new properties on config (in this case `join_distribution_type` set for nft_polygon.transfers), do I need to define them anywhere else?

File: dbt_project.yml
Comment (jkylling) [inline dbt_project.yml:53]: We don't need a fallback as we never use it.
```suggestion
      - sql: "{{ set_trino_session_property(is_join_distribution_enforced(model), 'join_distribution_type', model.config.get('join_distribution_type')) }}"
```

File: macros/set_trino_session_property.sql
Comment (belen-pruvost) [reply macros/set_trino_session_property.sql:13]: We can, yes 👍 I'll change this

**Comment by belen-pruvost** [comment]
Besides the fallback feedback, which I'll address, no property is being set on the model currently so I believe the new model property is perhaps not correctly configured. Any help is appreciated 😅 
Screenshot 2023-08-10 at 16 27 48

**Comment by belen-pruvost** [comment]
I've tested this locally and the session property is correctly set for nft_polygon_transfers (I added some logs to prove it)
Screenshot 2023-08-11 at 15 20 13

The CI also works fine  (by commenting out the check tags step). I will remove that `do not merge` commit

However, since this removes a prehook it implies an update on all dbt models. Not sure if there is something that needs to be done before merging the PR.

@jeff-dude + @couralex6 let me know what you think 🙏

**Review by aalan3 — APPROVED** [review]

**Comment by belen-pruvost** [comment]
Nice - thanks @aalan3 🙌

**Comment by jeff-dude** [comment]
this approach works for me. do we have others we want to apply to now too? there is a solana spell under orca_whirpool that we focused on last time too

**Comment by jeff-dude** [comment]
let's pay attention to #4055 here too (merge timing and when to change pre-hook there)

**Comment by belen-pruvost** [comment]
> this approach works for me. do we have others we want to apply to now too? there is a solana spell under orca_whirpool that we focused on last time too

Good call out - I'll add the prehook to this model too. Regarding https://github.com/duneanalytics/spellbook/pull/4055 I think we could go in sequence:
1) Merge this PR
2) Rebase https://github.com/duneanalytics/spellbook/pull/4055
3) If we get an exceeded local memory error on the sharky models, add the prehook

**Review by jeff-dude — APPROVED** [review]
sounds good, i'll look to prep prod for this merge and get it going

**Pull request merged**
